src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/BaseGraph.java

```diff
@@ -225,8 +225,8 @@ public final V getNextReferenceVertex( final V v, final boolean allowNonRefPaths
         final Set<E> blacklistedEdgeSet = blacklistedEdge.isPresent() ? Collections.singleton(blacklistedEdge.get()) : Collections.emptySet();
 
         // if we got here, then we aren't on a reference path
-        final Optional<E> edge = outgoingEdges.stream().filter(e -> !blacklistedEdgeSet.contains(e)).findAny();
-        return edge.isPresent() ? getEdgeTarget(edge.get()) : null;
+        final List<E> edges = outgoingEdges.stream().filter(e -> !blacklistedEdgeSet.contains(e)).limit(2).collect(Collectors.toList());
+        return edges.size() == 1 ? getEdgeTarget(edges.get(0)) : null;
```

Comment:
Getting the first edge is still essentially random in that it depends on the order in which reads arrived.  Why not be principled and traverse the edge with the greatest weight, which seems like a good heuristic if you're trying to arrive back at the reference?
```
final Optional<E> edge = outgoingEdges.stream().max(Comparator.comparingInt(BaseEdge::getMultiplicity).reversed())
```

Reply:
Sorry, I misread the code.  But I still think my suggestion to do a greedy search before giving up on finding the reference is reasonable and more robust to things like high depth where you might have lots of low-weight non-reference edges.

Reply:
Right, the current behavior is to pick a random edge. The PR reverts to the 3.x behavior, which is to return a non-reference edge iff there 1 non-reference edge.

It would be great to try a more exhaustive search. However, my intuition is that the we are unlikely to make it back to the reference when the paths start forking and the computational cost can be substantial. We can use this PR to revert back to the deterministic logic and save implementation (and evaluation) of a greedy search for a future issue.

Reply:
@davidbenjamin thoughts?

Reply:
@DonFreed I'm fine with the 3.x behavior.  Anything fancier can wait for a later thorough revision of the dangling end merging code.